chain/bitcoind_client.go

```diff
@@ -1132,68 +1077,71 @@ func (c *BitcoindClient) filterTx(tx *wire.MsgTx,
-	// can shortcut the filter process and let the caller know the filter
-	// matches.
-	c.clientMtx.RLock()
-	if _, ok := c.memPool[tx.TxHash()]; ok {
-		c.clientMtx.RUnlock()
+	// If we've already seen this transaction and it's now been confirmed,
+	// then we'll shortcut the filter process by immediately sending a
+	// notification to the caller that the filter matches.
+	if _, ok := c.mempool[tx.TxHash()]; ok {
 		if notify && blockDetails != nil {
 			c.onRelevantTx(rec, blockDetails)
 		}
 		return true, rec, nil
 	}
-	c.clientMtx.RUnlock()
 
-	// Cycle through outputs and check if we've matched a known address.
-	// Add any matched outpoints to watchOutPoints.
+	// Otherwise, this is a new transaction we have yet to see. We'll need
+	// to determine if this transaction is somehow relevant to the caller.
+	var isRelevant bool
+
+	// We'll start by cycling through its outputs to determine if it pays to
+	// any of the currently watched addresses. If an output matches, we'll
+	// add it to our watch list.
 	for i, out := range tx.TxOut {
 		_, addrs, _, err := txscript.ExtractPkScriptAddrs(
 			out.PkScript, c.chainParams)
 		if err != nil {
-			log.Debugf("Couldn't parse output script in %s:%d: %v",
+			log.Debugf("Unable to parse output script in %s:%d: %v",
 				tx.TxHash(), i, err)
 			continue
 		}
+
 		for _, addr := range addrs {
-			c.clientMtx.RLock()
-			if _, ok := c.watchAddrs[addr.EncodeAddress()]; ok {
-				notifyTx = true
-				c.watchOutPoints[wire.OutPoint{
+			if _, ok := c.watchedAddresses[addr.String()]; ok {
+				isRelevant = true
+				op := wire.OutPoint{
 					Hash:  tx.TxHash(),
 					Index: uint32(i),
-				}] = struct{}{}
+				}
+				c.watchedOutPoints[op] = struct{}{}
 			}
-			c.clientMtx.RUnlock()
 		}
 	}
 
-	// If an output hasn't already matched, see if an input will.
-	if !notifyTx {
+	// If the transaction didn't pay to any of our watched addresses, we'll
+	// check if it spends any of our watched outpoints.
+	if !isRelevant {
 		for _, in := range tx.TxIn {
-			c.clientMtx.RLock()
-			if _, ok := c.watchOutPoints[in.PreviousOutPoint]; ok {
-				c.clientMtx.RUnlock()
-				notifyTx = true
+			if _, ok := c.watchedOutPoints[in.PreviousOutPoint]; ok {
+				isRelevant = true
 				break
 			}
-			c.clientMtx.RUnlock()
 		}
 	}
 
-	// If we have a match and it's not mined, notify the TX. If the TX is
-	// mined, we notify as part of FilteredBlockConnected. The boolean map
-	// value will let us know if we last saw it as mined or unmined.
-	if notifyTx {
-		c.clientMtx.Lock()
-		if _, ok := c.memPool[tx.TxHash()]; blockDetails == nil || !ok {
-			c.onRelevantTx(rec, blockDetails)
-		}
-		c.memPool[tx.TxHash()] = struct{}{}
-		c.clientMtx.Unlock()
+	if !isRelevant {
+		return false, rec, nil
+	}
+
+	// If this transaction matched any of our filters but it has yet to be
+	// confirmed, we'll defer sending its notification until it has been
+	// mined as part of FilteredBlockConnected. Otherwise, we'll send a
```

Comment:
"Otherwise, we'll send a notification immediately" -> I don't see that this is being done